Source/ProgressionSystemRuntime/Private/Widgets/PSOverlayWidget.cpp

```diff
@@ -4,13 +4,119 @@
 #include "Widgets/PSOverlayWidget.h"
 
 #include "Components/Image.h"
+#include "Curves/CurveFloat.h"
+#include "Data/PSDataAsset.h"
 
 #include UE_INLINE_GENERATED_CPP_BY_NAME(PSOverlayWidget)
 
-// Sets the visibility of the background overlay and lock icon.
-void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate)
+// Sets the visibility of the overlay elements and playing fade animation if needed
+void UPSOverlayWidget::SetOverlayVisibility(ESlateVisibility VisibilitySlate, bool bShouldPlayFadeAnimation/* = false*/)
+{
+	if (!bShouldPlayFadeAnimation)
+	{
+		SetOverlayItemsVisibility(VisibilitySlate);
+		return;
+	}
+
+	bShouldPlayFadeAnimationInternal = bShouldPlayFadeAnimation;
+
+	if (VisibilitySlate == ESlateVisibility::Visible)
+	{
+		if (!ensureMsgf(PSCBackgroundOverlay, TEXT("ASSERT: [%i] %s:\n'PSCBackgroundOverlay' is not valid!"), __LINE__, *FString(__FUNCTION__)))
+		{
+			return;
+		}
+		if (!ensureMsgf(PSCBackgroundIconLock, TEXT("ASSERT: [%i] %s:\n'PSCBackgroundIconLock' is not valid!"), __LINE__, *FString(__FUNCTION__)))
+		{
+			return;
+		}
+		ESlateVisibility PrevOverlayOpacity = PSCOverlay->GetVisibility();
+
+		//if new visibility is same as previous animation is not required 
+		if (PrevOverlayOpacity == VisibilitySlate)
+		{
+			bShouldPlayFadeAnimationInternal = false;
+		}
+		bIsFadeInAnimationInternal = true;
+		SetOverlayItemsVisibility(VisibilitySlate);
+	}
+	else
+	{
+		bIsFadeInAnimationInternal = false;
+	}
+
+	const UWorld* World = GetWorld();
+	if (!ensureMsgf(World, TEXT("ASSERT: [%i] %s:\n'World' is not valid!"), __LINE__, *FString(__FUNCTION__)))
+	{
+		return;
+	}
+
+	StartTimeFadeAnimationInternal = World->GetTimeSeconds();
+}
+
+// overrides NativeTick to make the user widget tickable
+void UPSOverlayWidget::NativeTick(const FGeometry& MyGeometry, float InDeltaTime)
+{
+	Super::NativeTick(MyGeometry, InDeltaTime);
+
+	TickPlayFadeOverlayAnimation();
+}
+
+// Event to execute when widget is ready
+void UPSOverlayWidget::NativeConstruct()
+{
+	Super::NativeConstruct();
+}
+
+// Play the overlay elements fade-in/fade-out animation. Uses the internal FadeCurveFloatInternal initialized in NativeConstruct
+void UPSOverlayWidget::TickPlayFadeOverlayAnimation()
+{
+	if (!UPSDataAsset::Get().GetFadeCurveFloat() || !bShouldPlayFadeAnimationInternal)
+	{
+		return;
+	}
+
+	const UWorld* World = GetWorld();
+	if (!ensureMsgf(World, TEXT("ASSERT: 'World' is not valid")))
+	{
+		return;
+	}
+
+	const float SecondsSinceStart = GetWorld()->GetTimeSeconds() - StartTimeFadeAnimationInternal;
+	float OpacityValue = UPSDataAsset::Get().GetFadeCurveFloat()->GetFloatValue(SecondsSinceStart);
+
+	float MinTime = 0.f;
+	float MaxTime = 0.f;
+	if (!ensureMsgf(UPSDataAsset::Get().GetFadeCurveFloat(), TEXT("ASSERT: 'CurveFloat' is not valid")))
```

Comment:
The current code overcomplicates things by using a curve for what seems to be a simple linear fade animation. 

A float value for the duration in seconds of the fade is enough to handle this. We can simplify the logic by using a straightforward time-based linear interpolation without relying on a curve.

### Expected result
Next code is example, is not tested, but you can see general approach, I use here only next properties: bShouldPlayFadeAnimationInternal, StartTimeFadeAnimationInternal and FadeDuration.
Removed properties: `bIsFadeInAnimationInternal` and float curve

```cpp
void UPSOverlayWidget::TickPlayFadeOverlayAnimation()
{
	const UWorld* World = GetWorld();
	if (!bShouldPlayFadeAnimationInternal 
	    || !World 
	    || !ensureMsgf(FadeDuration > 0.0f, TEXT("ASSERT: [%i] %hs:\n'FadeDuration' must be greater than 0"), __LINE__, __FUNCTION__))
	{
		bShouldPlayFadeAnimationInternal = false;
		return;
	}

	const float SecondsSinceStart = World->GetTimeSeconds() - StartTimeFadeAnimationInternal;
	const float OpacityValue = FMath::Clamp(SecondsSinceStart / FadeDuration, 0.0f, 1.0f);

	if (SecondsSinceStart >= FadeDuration)
	{
		SetOverlayItemsVisibility(ESlateVisibility::Collapsed);
		bShouldPlayFadeAnimationInternal = false;
		return;
	}

	if (PSCOverlay)
	{
		PSCOverlay->SetRenderOpacity(OpacityValue);
	}
}
```

Reply:
removed curve float as part of: https://github.com/h4rdmol/ProgressionSystem/pull/20/commits/b31977e4e0c4f1a106d0c5d359e60d669bc2e155

Reply:
Please never expose any tweakable properties for designers right in blueprints, but only in data assets/tables according to data-driven approach.

Please move FadeDuration to the Data Asset

Reply:
done: https://github.com/h4rdmol/ProgressionSystem/pull/17/commits/1469e2393bcd04796da5e3dae5d4ee9b149ebe9e